### PR TITLE
Continue work on high-level API

### PR DESCRIPTION
**Pull request by maleadt**
Together with https://github.com/JuliaTelecom/SoapySDR.jl/pull/78, the following seems to make sense now (at least when inspecting the data we get back on the loopback; didn't test actualy Tx/Rx):

`simple_loopback_multichannel.jl`
```julia
ENV["SOAPY_SDR_LOG_LEVEL"] = "DEBUG"

using SoapySDR, Printf, Unitful

#SoapySDR.register_log_handler()

function loopback_test(num_channels=2)
    # open the first device
    devs = Devices(parse(KWArgs, "driver=XTRX"))
    dev = Device(devs[1])

    # get the RX and TX channels
    chan_rx = SoapySDR.Channel[]
    chan_tx = SoapySDR.Channel[]
    for i in 1:num_channels
        push!(chan_rx, dev.rx[i])
        push!(chan_tx, dev.tx[i])
    end

    # enable a loopback
    SoapySDR.SoapySDRDevice_writeSetting(dev, "LOOPBACK_ENABLE", "TRUE")

    # open RX and TX streams
    format = Complex{Int16} #chan_rx.native_stream_format
    stream_rx = SoapySDR.Stream(format, chan_rx)
    stream_tx = SoapySDR.Stream(format, chan_tx)

    # properties that matter for the buffers we'll be using
    mtu = Int(stream_tx.mtu)

    try
        wr_cnt = 0
        rd_cnt = 0
        should_transmit() = wr_cnt < 999
        should_receive() = rd_cnt < 3000
        # NOTE: we read more buffers in case there's delay between TX and RX

        SoapySDR.activate!(stream_rx)
        SoapySDR.activate!(stream_tx)

        while should_transmit() || should_receive()
            if should_transmit()
                buffs = []
                for i in 1:num_channels
                    buf = Vector{format}(undef, mtu)
                    buf .= (2*num_channels*wr_cnt+1+(i-1)*num_channels)+(2*num_channels*wr_cnt+2+(i-1)*num_channels)im
                    push!(buffs, buf)
                end
                write(stream_tx, tuple(buffs...))
                wr_cnt += 1
            end

            if should_receive()
                buffs = []
                for i in 1:num_channels
                    buf = Vector{format}(undef, mtu)
                    push!(buffs, buf)
                end
                read!(stream_rx, tuple(buffs...); timeout=5u"s")
                rd_cnt += 1

                if rd_cnt % 32 == 0
                    println()
                end
                @printf("%3d ", real(buffs[1][1]))
            end
        end
        println()
    finally
        SoapySDR.deactivate!(stream_tx)
        SoapySDR.deactivate!(stream_rx)

        # close everything
        finalize.([stream_rx, stream_tx])
        finalize(dev)
    end
end

isinteractive() || loopback_test()
isinteractive() || loopback_test()
```

### CODE REVIEW

File: software/soapysdr-xtrx/Streaming.cpp
Comment (staticfloat) [inline software/soapysdr-xtrx/Streaming.cpp:325]: So this basically says "couldn't transmit, and I'm zooming backward so that we have 255 buffers of space between the read pointer and the write pointer", right?  If that's the case, why can't we write into a buffer here?  Is it just so that we can surface errors better?

**Review by staticfloat — APPROVED** [review]
This looks pretty good to me, I think a test we should do before merging is to get rid of the [`soapy_{read,write}()` shims in libsigflow](https://github.com/JuliaComputing/xtrx_julia/blob/ab2d2d82614744e474a7411ab97a6b642274d6e5/software/scripts/libsigflow.jl#L73-L169), then run `acquire_iq.jl --full` to ensure that it works with all of our loopbacks and other code.  I'm pretty confident that it will, but that would be a nice regression test.

**Comment by sjkelly** [comment]
Let's not merge this yet. I would like to port over test_pattern so we can validate buffer ordering, but there seem to still be some issues I need to qualify.

File: software/soapysdr-xtrx/Streaming.cpp
Comment (staticfloat) [inline software/soapysdr-xtrx/Streaming.cpp:363]:
```suggestion
            dst_cs16[2*i + 1] = src_cs16[4*i + channel + 1];
```
I'm pretty sure we want `+1` here, since the data format is `Ai Aq Bi Bq` where `Ai` refers to the real part of channel A.  You can verify this by opening up LimeSuiteGUI, going to `LimeLight & PAD` tab, going to the `Sample position & Clock` sub-tab, and looking at the sample position configuration.

File: software/soapysdr-xtrx/Streaming.cpp
Comment (staticfloat) [inline software/soapysdr-xtrx/Streaming.cpp:380]:
```suggestion
            dst_cs16[4*i + channel + 1] = src_cs16[2*i + 1];
```

File: software/soapysdr-xtrx/Streaming.cpp
Comment (maleadt) [reply software/soapysdr-xtrx/Streaming.cpp:363]: I followed what Steve said here: https://github.com/JuliaComputing/xtrx_julia/pull/59#issue-1359387234

Comment (staticfloat) [reply software/soapysdr-xtrx/Streaming.cpp]: Yep, sorry, we disproved that comment but never wrote it down anywhere.

**Comment by sjkelly** [comment]
I guess we ought to sign extend in here as well while we are at it for simplicity. We can keep the old behavior by specify an unsigned stream type.

**Comment by staticfloat** [comment]
I think if we are going to sign-extend it should be in the FPGA.  If we ask for 12-bit samples, the FPGA should not sign-extend (obviously) but this is, IMO, a classic case of something good to do in the FPGA.

**Comment by sjkelly** [comment]
So there is a conflation here between the number of samples for the DMA API and the number of samples in the stream API. The requested samples are per channel in the stream API. This is causing some issues in my validation when interfacing with the higher level Soapy layers that rely on this to e.g. collect 1e6 samples in a single array.

**Comment by sjkelly** [comment]
Okay this should be good to go. Note that the acquire buffer section will need some more work, so I removed it for now: https://github.com/JuliaComputing/xtrx_julia/pull/89/commits/b5ed3376d2c8b3fd61bca3eb2b42b4058f2d1c8a

image

File: software/soapysdr-xtrx/Streaming.cpp
Comment (sjkelly) [reply software/soapysdr-xtrx/Streaming.cpp]: This logic needs to be revisited. I have removed it for now. We can add it back later.

**Comment by sjkelly** [comment]
I am merging this since I need it for several other tests and bringups, and as it stand this PR only adds the high level API. The change in https://github.com/JuliaComputing/xtrx_julia/commit/b5ed3376d2c8b3fd61bca3eb2b42b4058f2d1c8a
is related to the buffer handling, which we will need to revisit with synchronization as well. We can move that to a separate PR.

**Pull request merged**
